File: Data Structures & Algorithms/binary-tree-right-side-view/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
    using vi = vector<int>;
    using pi = pair<int, int>;

   public:
    vector<int> rightSideView(TreeNode* root) {
        if (!root) return {};  // Handle empty tree

        vi result;
        queue<TreeNode*> q;
        q.push(root);

        while (!q.empty()) {
            int levelSize = q.size();  // Number of nodes at the current level

            for (int i = 1; i <= levelSize; i++) {
                auto node = q.front();
                q.pop();

                if (i == levelSize) {
                    result.push_back(node->val);
                }

                // Push children to the queue for the next level
                if (node->left) q.push(node->left);
                if (node->right) q.push(node->right);
            }
        }

        return result;
    }
};

File: Data Structures & Algorithms/binary-tree-postorder-traversal/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<Integer> postorderTraversal(TreeNode root) {
        var result = new ArrayList<Integer>();
        postorderTraversal(root, result);
        return result;
    }

    private void postorderTraversal(TreeNode root, ArrayList<Integer> result) {
        // base case
        if (root == null)
            return;

        postorderTraversal(root.left, result);
        postorderTraversal(root.right, result);
        result.add(root.val);
    }
}